### DISCUSSION TITLE
How to globally access the JWT token from the request

### DISCUSSION THREAD

**Discussion by Feijo** [fastapi/fastapi#7082] [Questions]
I'm using a JWT authentication implementation as per the documentation: https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/

I can easily access the JWT token using DI in all my routers:

```
@router.get('/something')
async def some_function(token: str = Depends(oauth2_scheme)):
    # Some logic here
```

I'm trying to use SQLAlchemy's `before_compile` event to append some logic to all queries:

```
@event.listens_for(Query, 'before_compile', retval=True)
def sqlalchemy_before_compile(query: Query):
    # From here, how do I access the JWT token?
```

**Comment by Feijo** [comment]
As a workaround I ended up going with [starlette-context](https://github.com/tomwojcik/starlette-context) + custom plugin.

Just wondering if there's another solution?

**Comment by valentin994** [comment]
Do you mean accessing the JWT that the client sends you in his request?
If so you can add ```Authorize: AuthJWT = Depends()``` in your method, and 
then you just call ```Authorize.get_jwt_subject()```

**Comment by Feijo** [comment]
Like this you mean?

```
@event.listens_for(Query, 'before_compile', retval=True)
def sqlalchemy_before_compile(query: Query, authorize=Depends()) -> Query:
    jwt = authorize.get_jwt_subject()
```

**`AttributeError: 'Depends' object has no attribute 'get_jwt_subject'`**

Bare in mind that this is not a FastAPI route, this is a SQLAlchemy event.

**Comment by valentin994** [comment]
I meant in the route like this.

```
@router.get('/something')
async def some_function(Authorize: AuthJWT = Depends()):
    Authorize.jwt_required()
    what_you_used_to_create_the_token = Authorize.get_jwt_subject()
    function_to_use_that_subject(what_you_used_to_create_the_token)
```

So this looks if there is an access token cookie, if there is one the get_jwt_subject() will decode it back from what was used for the cookie to be made. For example, if you have a user register and login you would choose one unique identifier, let's say email, and then create a jwt with it, with this "something" route you would take the created cookies and get the email and then proceed to make database calls. Hope I didn't miss the point or your intent with the jwt.

**Comment by Feijo** [comment]
Hi Valentin, thanks for trying to help but I'm afraid I'm trying to do something a bit unrelated.

I'm not trying to access the raw JWT token in one of my routes. There's a SQLAlchemy event I'm capturing, this event fires before a query compiles internally by sqlalchemy. The handler for this event is not in a request context, I cannot using `Depends`. I'm yes trying to capture the JWT, but the event handler has nothing to do with routes.

That's fine, as I said, starlette-context provides access to the current request from a global scope.

**Comment by tiangolo** [comment]
Thanks for the help here @valentin994 ! :clap: :bow: 

Thanks for reporting back and closing the issue @Feijo  :+1:
